File: permutationsii.js
//Objective is the same as 'Permutations', but this time to only return 
//unique permutations.

let nums = [1,1,2]


//O(N!/(N-k)!) solution that is the same as 'Permutations' but this time we sort
//the array first to skip any of the same elements

nums.sort((a,b) => {
    return a - b
})

let temp = []
let result = []

backtrack(nums, temp)

function backtrack(nums, temp) {
    if (nums.length == 0) {
        result.push([...temp])
        return
    }

    for (let i = 0; i < nums.length; i++){
        if (i !== 0 && nums[i] == nums[i - 1]) {
            continue
        }
        temp.push(nums[i])
        nums.splice(i, 1)
        backtrack(nums, temp)
        nums.splice(i, 0, temp.pop())
    }
}

console.log(result)

File: permutationsii.java
//Java Solution

class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        LinkedList<Integer> temp = new LinkedList<>();
        
        List<Integer> nums2 = new ArrayList<>();
        for (int num : nums) {
            nums2.add(num);
        }
        
        backtrack(result, temp, nums2);
        return result;
    }
    
    public void backtrack(List<List<Integer>> result, LinkedList<Integer> temp, List<Integer> nums) {
        if (nums.size() == 0) {
            result.add(new ArrayList<>(temp));
            return;
        }
        
        for (int i = 0; i < nums.size(); i++) {
            if (i != 0 && nums.get(i) == nums.get(i - 1)) {
                continue;
            }
            temp.add(nums.get(i));
            nums.remove(i);
            backtrack(result, temp, nums);
            int last = temp.removeLast();
            nums.add(i, last);
        }
    }
}